lib/new_post_manager.rb

```diff
@@ -161,6 +161,26 @@ def self.default_handler(manager)
         result.errors.add(:base, I18n.t("js.errors.reasons.forbidden"))
         return result
       end
+
+      # we do not want people to be able to submit topics to the review
+      # queue without them tagging it first, if the category is set up
+      # to require tags
```

Comment:
why no call TopicCreator to ask it to do the validation. 

Maybe make https://github.com/discourse/discourse/blob/32689573fa7dc3389378b8ef669a01271438d71b/lib/topic_creator.rb#L167-L170 a bit more reusable?

Reply:
The logic there is a bit bizarre...if the tags are blank then it does the DiscourseTagging validation?? It should be doing that if the tags exist.

```ruby
    if @opts[:tags].blank?
      unless @guardian.is_staff? || !guardian.can_tag?(topic)
        category = find_category

        if !DiscourseTagging.validate_min_required_tags_for_category(@guardian, topic, category) ||
            !DiscourseTagging.validate_required_tags_from_group(@guardian, topic, category)
          rollback_from_errors!(topic)
        end
      end
    else
```

If the TopicCreator was doing the right thing here I think that we would achieve the same result, will try it.

Reply:
setup_tags isn't actually called until later; only `.valid?` for `TopicCreator` is called from `NewPostManager`. That in turn only calls `topic.valid?` which is not yet aware of the tags.

Do you think I should just move the tag validation checks from `NewPostManager` and `TopicCreator#create_tags` into `TopicCreator#valid?` I think that would be the best course of action but it does clutter up that method a bit. We cannot assign tags to a topic without it being created because `topic.tags=` goes off and creates the associations.

Tricky...

Reply:
`setup_tags` is the only other place in `TopicCreator` that does `errors.add(:base, whatever)` so it is an outlier. I am going to move this validation into `TopicCreator#valid?`